### ISSUE TITLE
Race condition between `delete_server` and convergence when deleting a group's lock znode

### ISSUE THREAD

**Issue by cyli** [rackerlabs/otter#1192]
See this error now when attempting to delete an empty scaling group:

```
Request failed: Unhandled Error: NotEmptyError((), {})", "otter_facility": "otter.rest.groups.delete_scaling_group", "request_status": "failed", "scaling_group_id": "1abcd010-5226-4cd0-b126-54ad317f1b08", "tenant_id": "000000", "traceback": "Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 524, in __bootstrap
    self.__bootstrap_inner()
  File "/usr/lib/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 504, in run
    self.__target(*self.__args, **self.__kwargs)
--- <exception caught here> ---
  File "/opt/otter/.ve/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 196, in _worker
    result = context.call(ctx, function, *args, **kwargs)
  File "/opt/otter/.ve/local/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/otter/.ve/local/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/opt/otter/.ve/local/lib/python2.7/site-packages/kazoo/client.py", line 1243, in delete
    return self._delete_recursive(path)
  File "/opt/otter/.ve/local/lib/python2.7/site-packages/kazoo/client.py", line 1278, in _delete_recursive
    self.delete(path)
  File "/opt/otter/.ve/local/lib/python2.7/site-packages/kazoo/client.py", line 1245, in delete
    return self.delete_async(path, version).get()
  File "/opt/otter/.ve/local/lib/python2.7/site-packages/kazoo/handlers/threading.py", line 102, in get
    raise self._exception
kazoo.exceptions.NotEmptyError: ((), {})
```

(Seen as a failure in the functional tests, `scaling_group.test_scaling_group_negative.ScalingGroupNegative.test_update_group_after_deletion`, line 239, when it's attempting to delete the empty group, before it attempts to update the group)

**Comment by cyli** [comment]
Also seen in the test `scaling_group.test_scaling_group_negative.ScalingGroupNegative.test_get_group_after_deletion`, line 216, when attempting to delete an empty group before getting the group.

**Comment by cyli** [comment]
This error seems reproducible on the dev vm, but it's intermittent.  When running the full functional test suite, it pretty much always happens, but when just running `test_delete_scaling_group`, sometimes it passes.

**Comment by cyli** [comment]
Ok, this seems to be caused by:

1.  Test creates a scaling group with `minEntities = 0`
1.  `create_scaling_group` calls `obey_launch_config` after creating the group, so that minEntities can be satisfied if it is there.  In the worker, if delta is 0 nothing happens, but for convergence a cycle is kicked off even if delta is 0 (see comment in code as to why that happens)
1.  Convergence cycle first gathers data, and then grabs a group lock to write to active servers.
1.  While convergence is gathering data, `delete_group` has acquired a lock on the group to delete all the cass data related to the group.
1.  `delete_group` releases the group lock.
1.  Convergence acquires the group lock in order to write the active servers.
1. `delete_group`'s callback attempts to delete the znode, which can't happen because convergence has acquired the lock.

The problem is the race condition with convergence and `delete_group` attempting to delete the node as soon as it releases the znode.

Possible solution:  deleting the znode is not integral to the behavior of deleting the group, it is just an optimization because extra znodes are left on Zookeeper and cause performance issues.  We can:

1. Attempt to delete the znode on every lock release, rather than on every group deletion - it already demonstrates that deleting fails if someone else has the lock
1. Fix kazoo so that it cleans up the znodes.

**Comment by manishtomar** [comment]
After implementing #1179 I am assuming that group will not be _really_ deleted after rest delete call. So, we can move the zk node delete from there to when converger will actually delete it.

**Comment by cyli** [comment]
Hmm... good point.  I was wondering what about if there was nothing to converge, since this only happens about 1/6 times when running the test singly, but I guess this is only a problem against mimic - on prod gathering the data would take so long this would probably definitely happen and the convergence cycle would happen after.

If we can fix it in kazoo though that might be nice, and a better place for it.  Do you know if that's practical or if there is a reason that it doesn't currently do so?

**Comment by manishtomar** [comment]
So, I don't think kazoo is to blame here since it's lock API is being incorrectly used. The contract is that calling lock.acquire and release in any combination from different client should be fine. But expecting kazoo to work when deleting internal znodes while acquiring or releasing lock (or even at other times) is not fair.

**Comment by radix** [comment]
@manishtomar - the "fix" that ying is talking about is to have kazoo delete the node itself so we wouldn't need to do the deletion ourselves. I'm gonna look into figuring out how we can clean up locks.

**Comment by radix** [comment]
so apparently the idea is to just repeatedly try to delete the node until it can successfully be deleted (i.e. when it has no child nodes). The Curator project has a tool for this called a ChildReaper, which just occasionally looks at all nodes in a lock directory and tries to delete them: https://github.com/apache/curator/blob/master/curator-recipes/src/main/java/org/apache/curator/framework/recipes/locks/Reaper.java#L257

**Comment by radix** [comment]
so after staring at the code a bit, I think the process is just this:

```
stat = exists(path)
if stat.getChildren() == 0:
    try:
        delete(path)
    except (NoNodeError, NotEmptyError):
        pass
```

That's *essentially* it, but the Curator daemon adds a bit of complexity by making sure the path has been empty for 3 consecutive checks before deleting it - presumably to reduce load on ZK.

**Comment by manishtomar** [comment]
In our case, I think the _ideal_ thing would be to have a reaper that goes through all the locks along with current groups in CASS and removes those are are not in CASS. But that the same time, I think deleting from converger should do the job if we can ensure that is the last entity requiring the lock.

**Comment by radix** [comment]
A reaper would be best, but a simple way to make our lock cleanup nicer in the short term would be to just 1) not recursively delete the lock (meaning that if it's in use, the deletion attempt will fail) and 2) retry deleting with some interval for a while, perhaps even several minutes.

**Comment by radix** [comment]
Closing this issue since the race condition is now pretty much fixed with #1298. #1179 is tracking the rest of the deletion-in-convergence work.
